checkpoint: improve HTTP error handling

ivan4th opened the pull request:
## Motivation

When `recovery.recovery-uri` is configured and an error happens, such as network glitch or a temporary server error due to overload, go-spacemesh continues its execution without recovery, which may not be always intended behavior. This, for example, causes `TestCheckpoint` systest flakes sometimes.
The way to fix it is to retry a few times upon transient errors (such as network glitch or a 503 server error), terminating after all the retries are exhausted, continue without an error upon `404 Not Found`, and terminate go-spacemesh in other cases.

## Description

When recovering from an URL, only continue without failing if checkpoint data request fails with 404 code. If the request fails with other 4xx HTTP error code, go-spacemesh startup fails. Upon other errors, including 5xx HTTP errors and network errors, retry checkpoint data request several times.`recovery.retry-max` and `recovery.retry-delay` specify retry parameters, defaulting to 5 and 3s, respectively. go-spacemesh terminates if all the retries fail.

## Test Plan

Systests should pass

ivan4th committed:
checkpoint: improve HTTP error handling

When recovering from an URL, only continue without failing if
checkpoint data request fails with 404 code. If the request fails
with other 4xx HTTP error code, go-spacemesh startup fails. Upon
other errors, including 5xx HTTP errors and network errors, retry
checkpoint data request several times.
`recovery.retry-count` and `recovery.retry-interval` specify retry
parameters, defaulting to 5 and 3s, respectively.

poszu commented on checkpoint/recovery.go:
```diff
@@ -89,11 +99,23 @@ func copyToLocalFile(
 	dst := RecoveryFilename(dataDir, filepath.Base(parsed.String()), restore)
-	if err = httpToLocalFile(ctx, parsed, fs, dst); err != nil {
-		return "", err
+	for range retryCount + 1 {
```
We already depend on and use https://github.com/hashicorp/go-retryablehttp. How about using it instead of implementing retries manually?

ivan4th replied:
Makes sense, I'll update the PR

ivan4th replied:
Fixed

ivan4th committed:
checkpoint: use retryablehttp for checkpoint data retrieval

ivan4th committed:
checkpoint: add --ignore-checkpoint-req-errors

When this flag is specified, go-spacemesh doesn't exit when recovery
URI request fails. Retries are still done, though.
This is needed b/c in the context of TestRecovery systest, all the
nodes are created with --recovery-uri flag passed, including the
bootnodes. This way, they can't successfully start unless the
bootstrapper is already running. But the bootstrapper requires an
working go-spacemesh node to start up, thus we get a circular
dependency.

fasmat commented on checkpoint/recovery.go:
```diff
@@ -170,7 +187,10 @@ func RecoverWithDb(
 	}
 	logger.Info("recover from uri", zap.String("uri", cfg.Uri))
-	cpFile, err := copyToLocalFile(ctx, logger, fs, cfg.DataDir, cfg.Uri, cfg.Restore)
+	cpFile, err := copyToLocalFile(
+		ctx, logger, fs, cfg.DataDir, cfg.Uri, cfg.Restore,
+		cfg.RetryMax, cfg.RetryDelay,
+	)
```
at that point it probably makes sense to change the function signature to

```suggestion
	cpFile, err := copyToLocalFile(ctx, logger, cfg, fs)
```

Wdyt?

ivan4th replied:
Fixed, thanks

fasmat commented on checkpoint/recovery_test.go:
```diff
@@ -119,33 +120,57 @@ func verifyDbContent(tb testing.TB, db *sql.Database) {
+	return checkpointServerWithFaultInjection(t, nil)
+}
+
 func TestRecover(t *testing.T) {
```
In my opinion we should keep tests much simpler than this and avoid sharing setup code with other tests that complicate them too much - generally as soon as an `if` pops up somewhere in a test we should consider writing it a new with less shared logic.

Compare the current implementation of `TestRecover` with this:

```go
func TestRecover(t *testing.T) {
	t.Parallel()

	setup := func(t *testing.T, handler http.HandlerFunc) (afero.Fs, *checkpoint.RecoverConfig) {
		t.Helper()
		fs := afero.NewMemMapFs()

		mux := http.NewServeMux()
		mux.HandleFunc("GET /snapshot-15", handler)
		ts := httptest.NewServer(mux)
		t.Cleanup(ts.Close)
		cfg := &checkpoint.RecoverConfig{
			GoldenAtx:   goldenAtx,
			DataDir:     t.TempDir(),
			DbFile:      "test.sql",
			LocalDbFile: "local.sql",
			NodeIDs:     []types.NodeID{types.RandomNodeID()},
			Uri:         fmt.Sprintf("%s/snapshot-15", ts.URL),
			Restore:     types.LayerID(recoverLayer),
			RetryMax:    5,
			RetryDelay:  100 * time.Millisecond,
		}
		require.NoError(t, fs.MkdirAll(filepath.Join(cfg.DataDir, bootstrap.DirName), 0o700))
		return fs, cfg
	}

	t.Run("http", func(t *testing.T) {
		db := sql.InMemory()
		localDB := localsql.InMemory()
		fs, cfg := setup(t, func(w http.ResponseWriter, r *http.Request) {
			w.Write([]byte(checkpointData))
		})

		data, err := checkpoint.RecoverWithDb(context.Background(), zaptest.NewLogger(t), db, localDB, fs, cfg)
		require.NoError(t, err)
		require.Nil(t, data)

		newDB, err := sql.Open("file:" + filepath.Join(cfg.DataDir, cfg.DbFile))
		require.NoError(t, err)
		require.NotNil(t, newDB)
		defer newDB.Close()
		verifyDbContent(t, newDB)

		restore, err := recovery.CheckpointInfo(newDB)
		require.NoError(t, err)
		require.EqualValues(t, recoverLayer, restore)
	})

	t.Run("http+retry", func(t *testing.T) {
		t.Parallel()
		db := sql.InMemory()
		localDB := localsql.InMemory()
		var fail atomic.Bool
		fail.Store(true)
		fs, cfg := setup(t, func(w http.ResponseWriter, r *http.Request) {
			if fail.CompareAndSwap(true, false) { // fail on first request
				w.WriteHeader(http.StatusServiceUnavailable)
				w.Write([]byte("service unavailable"))
				return
			}
			w.Write([]byte(checkpointData))
		})

		data, err := checkpoint.RecoverWithDb(context.Background(), zaptest.NewLogger(t), db, localDB, fs, cfg)
		require.NoError(t, err)
		require.Nil(t, data)

		newDB, err := sql.Open("file:" + filepath.Join(cfg.DataDir, cfg.DbFile))
		require.NoError(t, err)
		require.NotNil(t, newDB)
		defer newDB.Close()
		verifyDbContent(t, newDB)

		restore, err := recovery.CheckpointInfo(newDB)
		require.NoError(t, err)
		require.EqualValues(t, recoverLayer, restore)
	})

	t.Run("not found", func(t *testing.T) {
		t.Parallel()
		db := sql.InMemory()
		localDB := localsql.InMemory()
		fs, cfg := setup(t, func(w http.ResponseWriter, r *http.Request) {})
		cfg.Uri = strings.Replace(cfg.Uri, "/snapshot-15", "/snapshot-42", -1) // unavailable snapshot

		data, err := checkpoint.RecoverWithDb(context.Background(), zaptest.NewLogger(t), db, localDB, fs, cfg)
		require.ErrorIs(t, err, checkpoint.ErrCheckpointNotFound)
		require.Nil(t, data)
	})

	t.Run("url unreachable", func(t *testing.T) {
		t.Parallel()
		db := sql.InMemory()
		localDB := localsql.InMemory()
		fs, cfg := setup(t, func(w http.ResponseWriter, r *http.Request) {})
		cfg.Uri = "http://nowhere/snapshot-15" // unreachable url

		data, err := checkpoint.RecoverWithDb(context.Background(), zaptest.NewLogger(t), db, localDB, fs, cfg)
		require.ErrorIs(t, err, checkpoint.ErrCheckpointRequestFailed)
		require.Nil(t, data)
	})

	t.Run("unsupported scheme", func(t *testing.T) {
		t.Parallel()
		db := sql.InMemory()
		localDB := localsql.InMemory()
		fs, cfg := setup(t, func(w http.ResponseWriter, r *http.Request) {})
		cfg.Uri = "ftp://snapshot-15" // unsupported scheme

		data, err := checkpoint.RecoverWithDb(context.Background(), zaptest.NewLogger(t), db, localDB, fs, cfg)
		require.ErrorIs(t, err, checkpoint.ErrUrlSchemeNotSupported)
		require.Nil(t, data)
	})
}
```

It's in my opinion much clearer what is going on: each test is short, doesn't have any branches and doesn't share any code with tests outside of this specific group. This makes it easier to debug them and when they fail its easier to figure out why they failed instead of going through multiple indirections.

Wdyt?

acud replied:
I tend to agree with @fasmat. Ideally if the tests that check for failures start sharing the setup with the tests that test for happy path we end up with something which is less readable and more difficult to maintain. I would suggest to just separate the tests that depend on a mock returning an error to have a separate setup. IMO it is fine to duplicate simple code to get better readability.

ivan4th replied:
In general I think there needs some balance, as another thing to consider is that repeating code chunks also obscure the primary purpose of the test, which is in this case is "try to get the URL with or without problem => specific simple outcome"; this purpose being specified concisely in a table basically makes the purpose of the whole test more self-explanatory.
But in this particular case the tests do indeed differ enough from each other, so going with @fasmat's approach is reasonable enough, which I did.

fasmat commented on checkpoint/util.go:
```diff
@@ -109,22 +111,38 @@ func CopyFile(fs afero.Fs, src, dst string) error {
+
+	resp, err := c.Do(req)
+	if err != nil {
+		// This shouldn't really happen. According to net/http docs for Do:
+		// "Any returned error will be of type *url.Error."
+		return fmt.Errorf("%w: %w", ErrCheckpointRequestFailed, err)
```
I don't understand this comment. This does happen - we even have a test case for it, that asserts that this specific error is returned?

ivan4th replied:
Sorry, stray comment not removed from an earlier version of the change, will fix

ivan4th replied:
Stray comment removed

fasmat reviewed (commented):
I don't understand why we want to default to ignoring failing requests for checkpoints in system tests? Especially in the tests that do a checkpoint and restore these should be caught and logged and cause the test to fail early (failing to fetch the checkpoint) instead of having the restore "succeed" and then fail at assertion.

I think instead we need to make sure that the pods serving the checkpoints are up and available before starting the smeshing nodes. Wdys?

ivan4th commented:
> I don't understand why we want to default to ignoring failing requests for checkpoints in system tests? Especially in the tests that do a checkpoint and restore these should be caught and logged and cause the test to fail early (failing to fetch the checkpoint) instead of having the restore "succeed" and then fail at assertion.
> 
> I think instead we need to make sure that the pods serving the checkpoints are up and available before starting the smeshing nodes. Wdys?

The problem is more complicated than that, I explained it in the [commit message](https://github.com/spacemeshos/go-spacemesh/pull/6130/commits/7ee77a1465de2465e2bf62cf56575ba197600b40) but perhaps should comment in the code: 

This is needed b/c in the context of TestRecovery systest, all the nodes are created with --recovery-uri flag passed, including the bootnodes. This way, they can't successfully start unless the bootstrapper is already running. But the bootstrapper requires an working go-spacemesh node to start up, thus we get a circulardependency.

This circular dependency is rather hard to resolve, it involves avoiding bootstrapper restart but that's not easily doable b/c it will involve some substantial restructuring of systest code

ivan4th committed:
Address comments

ivan4th commented:
Added a comment explaining why go-spacemesh needs to continue after unsuccessfully retrying checkpoint recovery several times. We may want to fix this unfortunate circular dependency of node-bootstrapper startup in a followup.

ivan4th committed:
Fix typo
